### PR TITLE
HUB-174: Show hub resume page to paused users

### PR DESCRIPTION
**Pull request by chrisclayson**
Add resume page (similar to single IDP handover page)
Modify authn_request_controller to check cookie for pending status and redirect if present

### CODE REVIEW

File: spec/features/user_visits_resume_registration_page_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  before(:each) do
+    set_session_and_session_cookies!
+    stub_api_idp_list_for_sign_in
+    set_selected_idp_in_session(entity_id: 'http://idcorp.com', simple_id: 'stub-idp-one')
+    stub_translations
+  end
+
+  context 'and has a cookie containing a PENDING state and valid IDP identifiers' do
+    before(:each) do
+      # page.set_rack_session(transaction_simple_id: CONTINUE_ON_FAILED_REGISTRATION_RP)
```
Comment (jakubmiarka) [inline spec/features/user_visits_resume_registration_page_spec.rb]: leftover?

File: spec/features/user_visits_resume_registration_page_spec.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    end
+
+    it 'displays correct text and button' do
+      set_journey_hint_cookie('http://idcorp.com', 'PENDING', 'en')
```
Comment (jakubmiarka) [inline spec/features/user_visits_resume_registration_page_spec.rb]: the 'en' is not required, it already defaults to 'en' if not supplied

File: spec/features/user_visits_resume_registration_page_spec.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+      visit '/resume-registration'
+
+      expect(page).to have_content t('hub.paused_registration.resume.intro', rp_name: 'Test RP', display_name: 'IDCorp')
+      expect(page).to have_content t('hub.paused_registration.resume.heading', rp_name: 'Test RP', display_name: 'IDCorp')
+      expect(page).to have_button t('hub.paused_registration.resume.continue', display_name: 'IDCorp')
+      expect(page).to have_content t('hub.paused_registration.resume.alternative_other_ways', rp_name: 'Test RP')
```
Comment (jakubmiarka) [inline spec/features/user_visits_resume_registration_page_spec.rb]: I wonder if we should have another test which clicks the IDP?

File: app/controllers/authn_request_controller.rb [hunk 4]
```diff
@@ -67,6 +71,8 @@ def set_transaction_homepage(transaction_homepage)
   def redirect_for_journey_hint(hint)
     if !cookies.encrypted[CookieNames::VERIFY_SINGLE_IDP_JOURNEY].nil? && SINGLE_IDP_FEATURE
       redirect_to continue_to_your_idp_path
+    elsif !last_status.nil? && last_status['STATUS'] == 'PENDING' && hint != 'submission_confirmation'
```
[the remark this answers is not in the document]
Comment (jakubmiarka) [reply app/controllers/authn_request_controller.rb]: Then we could re-think this if logic, feels to complex. We might want to split it up a bit by single idp/journey-hint/eidas...

File: app/controllers/paused_registration_controller.rb [hunk 5]
```diff
@@ -1,24 +1,54 @@
...
+
+
 class PausedRegistrationController < ApplicationController
+  include JourneyHintingPartialController
+  include ViewableIdpPartialController
+
   # Validate the session manually within the action, as we don't want the normal 'no session' page.
-  skip_before_action :validate_session
+  skip_before_action :validate_session, except: :resume
   skip_before_action :set_piwik_custom_variables
```
[the remark this answers is not in the document]
Comment (chrisclayson) [reply app/controllers/paused_registration_controller.rb]: Good knowledge ... thanks!

File: app/models/feedback_source_mapper.rb [hunk 6]
```diff
@@ -46,7 +46,8 @@ def initialize(product_page_url)
       'PROOF_OF_ADDRESS' => 'select_proof_of_address',
       'CONFIRMING_IT_IS_YOU_PAGE' => 'confirming_it_is_you',
       'PROVE_IDENTITY_PAGE' => 'prove_identity',
-      'CONTINUE_TO_YOUR_IDP_PAGE' => 'continue_to_your_idp'
+      'CONTINUE_TO_YOUR_IDP_PAGE' => 'continue_to_your_idp',
+      'RESUME_PAGE' => 'resume_with_idp'
```
[the remark this answers is not in the document]
Comment (chrisclayson) [reply app/models/feedback_source_mapper.rb]: Good knowledge, thanks!

File: app/views/paused_registration/resume_with_idp.html.erb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+<div class="grid-row js-continue-to-idp" data-location="<%= url_for(controller: 'sign_in', action: 'select_idp_ajax', locale: I18n.locale)  %>">
+  <h1 class="heading-large"><%= t 'hub.paused_registration.resume.heading', display_name: @idp.display_name, rp_name: @transaction[:name] %></h1>
+
+  <%= t 'hub.paused_registration.resume.intro', display_name: @idp.display_name, rp_name: @transaction[:name] %>
+
+  <div class="actions">
+    <%= form_tag('', class: 'js-idp-form') do %>
+      <%= hidden_field_tag 'entity_id', @idp.entity_id, id: nil, class: 'js-entity-id' %>
+      <%= hidden_field_tag 'singleIdpJourneyIdentifier', @uuid %>
```
[the remark this answers is not in the document]
Comment (chrisclayson) [reply app/views/paused_registration/resume_with_idp.html.erb]: Didn't I put a similar comment on one of your PRs ... well, spotted!

[the remark this answers is not in the document]
Comment (jakubmiarka) [reply app/views/paused_registration/resume_with_idp.html.erb]: I never forget! :-D

File: spec/features/user_visits_resume_registration_page_spec.rb (hunk 1 above)
Comment (chrisclayson) [reply spec/features/user_visits_resume_registration_page_spec.rb]: Yes, forgot to remove.

File: app/views/paused_registration/resume_with_idp.html.erb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      <%= button_tag t('hub.paused_registration.resume.continue', display_name: @idp.display_name),
+                     class: 'button',
+                     id: 'continue-to-idp-button',
+                     type: 'submit'
+      %>
+    <% end %>
+    <br>
+    <%= t('hub.paused_registration.resume.alternatives') %>
+    <ul class="list-bullet list small-details">
+      <li><%= link_to t('hub.paused_registration.resume.alternative_new_identity'), start_path %></li>
```
[the remark this answers is not in the document]
Comment (chrisclayson) [reply app/views/paused_registration/resume_with_idp.html.erb]: Again ... good knowledge, thanks.

File: app/controllers/partials/journey_hinting_partial_controller.rb [hunk 9]
```diff
@@ -19,6 +19,17 @@ def last_status
...
   end
 
+  def is_last_status?(status)
+    last_status_value = last_status
+    !last_status_value.nil? && last_status_value['STATUS'] == status
+  end
+
+  def last_idp
+    last_status_value = last_status
+    return nil if last_status_value.nil? || last_status_value['IDP'].nil?
```
Comment (jakubmiarka) [inline app/controllers/partials/journey_hinting_partial_controller.rb]: do we need the second (`last_status_value['IDP'].nil?`) check? we could just return `last_status_value['IDP']` and it'll be either a value or nil.

File: config/locales/cy.yml [hunk 10]
```diff
@@ -551,6 +552,15 @@ cy:
             <li>bookmark this page</li>
           </ul>
           <p>You'll need the email or username and password you used to create an account.</p>
+      resume:
```
Comment (jakubmiarka) [inline config/locales/cy.yml]: the welsh translations now has come through, so we can update them

**Pull request merged**